libraries/rib-base/src/main/java/com/badoo/ribs/clienthelper/childawareness/ChildAwareRegistryImpl.kt

```diff
@@ -0,0 +1,88 @@
+package com.badoo.ribs.clienthelper.childawareness
+
+import androidx.lifecycle.Lifecycle
+import com.badoo.ribs.clienthelper.childawareness.ChildAwareCallbackInfo.Mode
+import com.badoo.ribs.core.Node
+import com.badoo.ribs.core.Rib
+import kotlin.reflect.KClass
+
+class ChildAwareRegistryImpl : ChildAwareRegistry {
+
+    private val callbacks: MutableList<ChildAwareCallbackInfo> = ArrayList()
+
+    override lateinit var node: Node<*>
+        private set
+
+    override fun init(node: Node<*>) {
+        this.node = node
+    }
+
+    override fun <T : Rib> whenChildBuilt(
+        lifecycle: Lifecycle,
+        child: KClass<T>,
+        callback: ChildCallback<T>
+    ) {
+        val info = ChildAwareCallbackInfo.Single(child, callback, lifecycle, Mode.ON_BUILT)
+        callbacks += info
+        notifyWhenRegistered(info)
+    }
+
+    override fun <T : Rib> whenChildAttached(
+        lifecycle: Lifecycle,
+        child: KClass<T>,
+        callback: ChildCallback<T>
+    ) {
+        val info = ChildAwareCallbackInfo.Single(child, callback, lifecycle, Mode.ON_ATTACH)
+        callbacks += info
+        notifyWhenRegistered(info)
+    }
+
+    override fun <T1 : Rib, T2 : Rib> whenChildrenBuilt(
+        lifecycle: Lifecycle,
+        child1: KClass<T1>,
+        child2: KClass<T2>,
+        callback: ChildrenCallback<T1, T2>
+    ) {
+        val info =
+            ChildAwareCallbackInfo.Double(child1, child2, callback, lifecycle, Mode.ON_BUILT)
+        callbacks += info
+        notifyWhenRegistered(info)
+    }
+
+    override fun <T : Rib, R : Rib> whenChildrenAttached(
+        lifecycle: Lifecycle,
+        child1: KClass<T>,
+        child2: KClass<R>,
+        callback: ChildrenCallback<T, R>,
+    ) {
+        val info =
+            ChildAwareCallbackInfo.Double(child1, child2, callback, lifecycle, Mode.ON_ATTACH)
+        callbacks += info
+        notifyWhenRegistered(info)
+    }
+
+    override fun onChildBuilt(child: Node<*>) {
+        notifyWhenChanged(Mode.ON_BUILT, child, node.children)
+    }
+
+    override fun onChildAttached(child: Node<*>) {
+        notifyWhenChanged(Mode.ON_ATTACH, child, node.children)
+    }
+
+    private fun notifyWhenRegistered(callback: ChildAwareCallbackInfo) {
+        when (callback) {
+            is ChildAwareCallbackInfo.Double<*, *> -> callback.invokeIfRequired(node.children)
```

Comment:
One thing we realised when talking with @antonshilov is that we need to pay attention to `node.children` not always containing what we would assume.

With portals, it will be the `PortalNode`'s `children` that will contain a remote child (for managing lifecycle and forwarding back press handling for example), but it's the virtual parent that will receive `onChildBuilt` callbacks (as it will want to set up connection / business logic). 

So in that case, the `node.children` in the virtual parent will not contain the expected remote child. To support this case with your solution, probably we need to introduce a second list, and use a distinguishing naming (e.g. `managedChildren` and `virtualChildren`).

Reply:
Portal support will be addressed in a separate PR.